bitchat/Features/media/ImageUtils.swift

```diff
@@ -0,0 +1,149 @@
+        }
+        let outputURL = try makeOutputURL()
+        try jpegData.write(to: outputURL, options: .atomic)
+        return outputURL
+    }
+
+    private static func scaledImage(_ image: UIImage, maxDimension: CGFloat) -> UIImage {
+        let size = image.size
+        let maxSide = max(size.width, size.height)
+        guard maxSide > maxDimension else { return image }
+        let scale = maxDimension / maxSide
+        let newSize = CGSize(width: size.width * scale, height: size.height * scale)
+        UIGraphicsBeginImageContextWithOptions(newSize, true, 1.0)
+        image.draw(in: CGRect(origin: .zero, size: newSize))
+        let rendered = UIGraphicsGetImageFromCurrentImageContext()
+        UIGraphicsEndImageContext()
+        return rendered ?? image
+    }
+    #else
+    static func processImage(_ image: NSImage, maxDimension: CGFloat = 512) throws -> URL {
+        let scaled = scaledImage(image, maxDimension: maxDimension)
+        guard let inputCG = scaled.cgImage(forProposedRect: nil, context: nil, hints: nil) else {
+            throw ImageUtilsError.encodingFailed
+        }
+        let width = inputCG.width
+        let height = inputCG.height
+        let colorSpace = CGColorSpace(name: CGColorSpace.sRGB) ?? CGColorSpaceCreateDeviceRGB()
+        guard let context = CGContext(
+            data: nil,
+            width: width,
+            height: height,
+            bitsPerComponent: 8,
+            bytesPerRow: 0,
+            space: colorSpace,
+            bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
+        ) else {
+            throw ImageUtilsError.encodingFailed
+        }
+        context.draw(inputCG, in: CGRect(x: 0, y: 0, width: width, height: height))
+        guard let cgImage = context.makeImage() else {
+            throw ImageUtilsError.encodingFailed
+        }
+        var quality = compressionQuality
+        guard var jpegData = encodeJPEG(from: cgImage, quality: quality) else {
+            throw ImageUtilsError.encodingFailed
+        }
+        while jpegData.count > targetImageBytes && quality > 0.3 {
+            quality -= 0.1
+            guard let next = encodeJPEG(from: cgImage, quality: quality) else { break }
+            jpegData = next
+        }
+        let outputURL = try makeOutputURL()
+        try jpegData.write(to: outputURL, options: .atomic)
+        return outputURL
+    }
+
+    private static func scaledImage(_ image: NSImage, maxDimension: CGFloat) -> NSImage {
+        let size = image.size
+        let maxSide = max(size.width, size.height)
+        guard maxSide > maxDimension else { return image }
+        let scale = maxDimension / maxSide
+        let newSize = NSSize(width: size.width * scale, height: size.height * scale)
+        let scaledImage = NSImage(size: newSize)
+        scaledImage.lockFocus()
+        image.draw(in: NSRect(origin: .zero, size: newSize),
+                   from: NSRect(origin: .zero, size: size),
```

Comment:
We could probably unify the scaling math to avoid duplicating the logic and avoiding any room for bugs